Re-check Tpackage scope escapes after normalising paths

mrmr1993 opened the pull request:
This PR uses the original unmarked level of `Tpackage`s when rechecking after path normalisation. Previously, this check did not run due to the marking check at the top of the function.

mrmr1993 committed:
Fix: Re-check Tpackage scope escapes after normalising paths

gasche commented:
For contributors less versed in the internals of the type-checker, more context would help, for example:
- I suppose this is related to the modular explicits work, providing a cross-reference would be helpful for context
- in the description, could you show a program where the change makes a difference, and explain why?
- ideally this program would also be included as a regression test in the patch

mrmr1993 commented:
I don't think this has any user-visible impact, and this doesn't directly relate to the modular explicits work.

However, the trunk behaviour of the `check_scope_escape` function was incorrect. For example, in this situation:
```ocaml
module type S = sig
  module type S1
  module type S2 = S1
end

let f (module M : S) =
  (* you are here *)
```
calling `check_scope_escape` will correctly identify whether that the type `(module M.S1)` escapes, but will not notice if `(module M.S2)` escapes. This is because calling  `check_scope_escape` on `(module M.S2)`
* checks that `ty.level >= lowest_level` (for `ty =` the original `(module M.S2)` type)
* modifies the type's level so that `ty.level < lowest_level`, so that it knows not to run the checks on it again
  - `ty.level <- pivot_level - ty.level` in the code (`pivot_level = 2 * lowest_level - 1`)
* notices that `M` would escape in `M.S2`, so it replaces it with its definition (`M.S1`) to see if that has a wider scope
* creates a copy of the original `(module M.S2)` type, modifies the copy's description to `(module M.S1)`, and calls itself with it
  - notably, the `(module M.S1)` type copies the level, which is `< lowest_level`. **This is the bug.**
* the recursive call checks that `ty.level >= lowest_level` (for `ty = ` the new `(module M.S1)` type). It's not, so it assumes that it's checked this type already and found no escape.

All this PR does is take the original level rather than the modified level in the copied type, so that the recursive call will run an escape check for the `(module M.S1)` type.

gasche commented on typing/ctype.ml:
```diff
@@ -820,7 +820,11 @@ let check_scope_escape env level ty =
+            { ty with
+              desc = Tpackage (p', nl, tl)
+              (* Use unmarked level, so that we check again. *)
+            ; level= pivot_level - ty.level }
```
I would find it clearer to save the pre-assignment `ty.level` value to a variable and use it here; `let unmarked_level = ty.level in`?

gasche reviewed (commented):
Thanks for the explanation!

In the `Tconstr` case, we also do a recursive call on `aux ty'` when expansion succeeds. Do we have the same problem, where the call could silently succeed because `ty'` incorrectly uses the same "marked" level? In that case, could it be possible to move the level marking below the per-type check, to avoid this issue (instead of being careful to restore the level when necessary)?

I would still prefer to see a regression test in the testsuite. Could you build an example program that was typed incorrectly before (I suppose: it was accepted when in fact the signature escapes?), and is now handled correctly (I suppose: it is rejected with an esacpe error)? Look for "* expect" tests in the typing-* test directories for inspiration.

gasche commented:
cc @trefis who introduced this code in 4a6f33d5577b5253609051406796b3aad1ededee as part of #2108. ( I had no memory of it of course, but I'm the one who reviewed that change, and it looks like we let bugs slip in. )

mrmr1993 commented:
> In the `Tconstr` case, we also do a recursive call on `aux ty'` when expansion succeeds. Do we have the same problem, where the call could silently succeed because `ty'` incorrectly uses the same "marked" level?

That should be fine, because `forward_try_expand_once` will set levels correctly.

> In that case, could it be possible to move the level marking below the per-type check, to avoid this issue (instead of being careful to restore the level when necessary)?

That makes sense. I've pushed a change to store and use the original error, but I'll do a proper rewrite now.

This whole function is a marking accident waiting to happen anyway: in an escape error is raised, types' levels aren't reset. In its current uses, the types that may be left marked (both those passed as arguments and those in the environment retrieved by the `Tconstr` case) are thrown away if an error occurs, but ideally this should be replaced with an implementation that doesn't leave types in a bad state. I'm going to rewrite using the `Btypes.snapshot`/`Btypes.backtrack`  implementation, which will do the right thing *and* decrease recursion depth.

> I would still prefer to see a regression test in the testsuite. Could you build an example program that was typed incorrectly before [...], and is now handled correctly [...]?

No. This function is only currently used in some cases when checking patterns (in `typecore.ml`), but the types that erroneously pass this check are always caught by a later unification (which calls `update_level`, where this check is implemented properly).

mrmr1993 committed:
Rewrite check_scope_escape using proper marking and unmarking

This uses the Btype.snapshot/backtrack mechanism, to ensure that we
always undo marking on types from the environment and to avoid a
`try ... with ...` construction for each recursive call.

gasche commented:
> That should be fine, because `forward_try_expand_once` will set levels correctly.

In the code of `expand_abbrev_gen`, I see (in one of the cases) a call to `update_level env level ty'`, where `level` is the level of the input (so here the *marked* level, if I read things correctly) and `ty'` is result of the expansion. If I read the code of `update_level` correctly, this should fail with a scope-escape exception (I'm assuming `level < ty.scope` will hold when `level` is below `lowest_level`), which is not what I would expect from expansion when it succeeds.

>  I'm going to rewrite using the `Btypes.snapshot/Btypes.backtrack` implementation, which will do the right thing and decrease recursion depth.

I've been looking at the code more, and I agree that it is hard to reason about. I wanted to try to delay marking the levels until after the type is normalized, but then I'm not sure that we cannot have non-terminating loops in the normalization logic. One think I like about using snapshot/backtrack is that you can mirror the way this is done in `update_level`, which does delay the level-marking logic, but uses `link_type` and `set_type_desc` during normalization, and we know we are as correct as this better-tested function.

mrmr1993 commented:
> In the code of `expand_abbrev_gen`, I see (in one of the cases) a call to `update_level env level ty'`, where `level` is the level of the input (so here the _marked_ level, if I read things correctly)

Apologies, you are right. As written, `check_scope_escape` effectively passes for any escaping path that has some expansion in the environment.

I have updated this with a proper rewrite of the function. I think it should now handle all of these cases correctly.

(Note: ~~at a glance, it seems like `occur_univar` may mark non-generic types from the environment and erroneously never unmark them, for similar reasons, so a similar rewrite is probably needed there too..~~ *Edit: On review, the environment types never get marked, so all is well.*)

gasche commented on typing/ctype.ml:
```diff
@@ -803,34 +803,45 @@ let rec normalize_package_path env p =
+        begin match !forward_try_expand_once env ty with
+        | ty' ->
+            mark ty;
+            check_scope_escape env level ty'
```
`update_level` calls `link_type` here instead of marking, is there a reason to not do the same thing? I am not completely sure that marking is correct (could it be that `ty'` is in fact `ty` ?), and I like the idea following the same traversal logic.

mrmr1993 replied:
> could it be that `ty'` is in fact `ty` ?

Unless something is very wrong elsewhere, `expand_abbrev_gen` will raise `Cannot_expand` if they would be the same. The `link_type` call only serves to replace the type with its expanded type, which we don't care about here because we will never re-visit this type before backtracking (due to marking).

gasche commented on typing/ctype.ml:
```diff
@@ -803,34 +803,45 @@ let rec normalize_package_path env p =
+        if Path.same p p' then raise Trace.(Unify [escape (Module_type p)]);
+        let orig_level = ty.level in
+        mark ty;
+        check_scope_escape env level
```
Same question as above: `update_level` uses `set_type_desc` in this case instead of marking, could we do the same thing?

mrmr1993 replied:
Same logic here: we need not worry about updating the type description because it will never get used before backtracking erases it.

gasche commented:
I like the new implementation better, but as someone who previously approved an incorrect version of the code I should probably get another pair of eyes. (cc @trefis @garrigue)

trefis commented on typing/ctype.ml:
```diff
@@ -803,34 +803,45 @@ let rec normalize_package_path env p =
+        let orig_level = ty.level in
+        mark ty;
+        check_scope_escape env level
+          {ty with desc= Tpackage (p', nl, tl); level= orig_level}
```
Actually, this `{ ty with ... }` just looks wrong, less so than before, but still wrong.
It seems better to just call `Btype.newty2`.

mrmr1993 replied:
Fixed, thanks.

trefis reviewed (changes requested):
I agree with the initial report: there is indeed a bug. I think the fix is still not quite right (using functional update to create new types seems like a bad idea ...): we're reusing the `id`, which could also be used to track types which have been visited (for instance in `lower_contravariant`) instead of the level marking nonsense. You could imagine this function being changed to use that one day, and the bug reappearing.

Apart from that, I like the additional calls to `snapshot` and `backtrack`.

mrmr1993 committed:
Use newty2 instead of record `with` syntax

trefis reviewed (approved):
Needs a Changes entry but good to merge otherwise.

mrmr1993 committed:
Add changes entry

mrmr1993 commented:
Woops, thought I'd already added one! Thanks @trefis.

mrmr1993 committed:
Merge branch 'trunk' into fix/package-scope-escape
